src/cpp/src/llm_pipeline.cpp

```diff
@@ -339,11 +400,354 @@ std::pair<std::string, Any> generation_config(const GenerationConfig& config) {
+    // (3) Reshape both models to static shape
+    // FIXME: There must be better logic than just hardcoded values
+    m_kvcache_desc = KVCacheDesc { 1024u, 0u };
+    const uint32_t max_prompt_size = m_kvcache_desc.total_size;
+    const uint32_t max_kvcache_size = m_kvcache_desc.total_size;
+    reshape_to_static(prefill_model, max_prompt_size, max_kvcache_size);
+    reshape_to_static(kvcache_model, 1u, max_kvcache_size);
+    // (4) Add slices to kvcache model
+    kvcache_model = add_slices_to_kvcache_inputs(kvcache_model);
+    // (5) Compiled both models for NPUW
+    std::map<std::string, std::string> cfg = { {"NPU_USE_NPUW", "YES"},
+                                               {"NPU_COMPILATION_MODE_PARAMS", "compute-layers-with-higher-precision=Sqrt,Power,ReduceMean,Add" },
+                                               {"NPUW_FOLD", "YES"},
+                                               {"NPUW_DCOFF_TYPE", "f16"},
+                                               {"NPUW_DCOFF_SCALE", "YES"},
+                                               {"NPUW_ONLINE_PIPELINE", "NONE"} };
+    ov::AnyMap properties{cfg.begin(), cfg.end()};
+    m_prefill_request = core.compile_model(prefill_model, "NPU", properties).create_infer_request();
+    m_kvcache_request = core.compile_model(kvcache_model, "NPU", properties).create_infer_request();
+    // (6) Initialize tensors
+    prepareForNewConversation();
+};
+
+NPULLMPipelineImpl::NPULLMPipelineImpl(
+    const std::filesystem::path& path,
+    const ov::AnyMap& config
+) : NPULLMPipelineImpl(path, path.string(), config) {
+}
+
+void NPULLMPipelineImpl::prepareForNewConversation() {
+    fill_tensor(m_prefill_request.get_tensor("input_ids"), m_tokenizer.get_pad_token_id());
+    fill_tensor(m_prefill_request.get_tensor("position_ids"), 0u);
+    fill_tensor(m_prefill_request.get_tensor("attention_mask"), 0u);
+    fill_tensor(m_kvcache_request.get_tensor("attention_mask"), 0u);
+    m_kvcache_desc.num_stored_tokens = 0u;
+}
+
+DecodedResults NPULLMPipelineImpl::generate(
+    StringInputs inputs,
+    OptionalGenerationConfig generation_config,
+    StreamerVariant streamer
+) {
+    GenerationConfig config = (generation_config.has_value()) ? *generation_config : m_generation_config;
+    if (std::holds_alternative<std::vector<std::string>>(inputs)) {
+        OPENVINO_THROW("Currently only batch size=1 is supported for NPU device");
+    }
+
+    OPENVINO_ASSERT(std::holds_alternative<std::string>(inputs));
+    const auto& text = std::get<std::string>(inputs);
+    auto tokenized_input = m_tokenizer.encode(std::get<std::string>(inputs));
+    remove_special_tokens(tokenized_input, text, {m_tokenizer.get_eos_token(), m_tokenizer.get_bos_token()});
+    auto encoded_results = generate(tokenized_input, config, streamer);
+    return {m_tokenizer.decode(encoded_results.tokens), encoded_results.scores};
+}
+
+EncodedResults NPULLMPipelineImpl::generate(
```

Comment:
can we move it to a dedicated file?

Reply:
Ideally we should, to keep it isolated. Most probably we will need additional `llm_pipeline_impl_base.hpp` (not to define `LLMPipelineImplBase` in public header) and `npu_pipeline_impl.hpp` with their corresponding `.cpp` files

Reply:
agreed to have it as a requirement for merging